test/runtests.jl

```diff
@@ -1,88 +1,163 @@
+            if (:dry_run => true) ∉ kw
+                push!(installed_versions, v_inst)
+            end
+        end
+    #end
+
+    commands = unique(vcat("julia", [["julia-$(v.major).$(v.minor)", "julia-$v"]
+        for v in vcat(UpdateJulia.nightly_version[], versions)]...))
+    @test String["julia", "julia-1.8", "julia-1.5.1", "julia-1.6.2", "julia-1.5.0-rc2",
+        "julia-1.7.0-rc1", "julia-1.7.0-beta2"] ⊊ commands
+    Sys.iswindows() || @test String["julia-1.0", "julia-1.4",  "julia-1.3.1",
+        "julia-1.3.0-rc3", "julia-1.3.0-alpha"] ⊊ commands
+
+
+    @testset "check" begin
+        for c in commands
+            installed = filter(x->startswith(string(x), c[7:end]), installed_versions)
+            actual = UpdateJulia.version_of(c)
+            for i in installed
+                if UpdateJulia.prefer(i, actual)
+                    println("Running: $c")
+                    println("Gives result: $actual")
+                    println("Installed these compatible versions, in this order: $installed")
+                    println("Specifically, $i, which is better than $actual")
+                    @test false
+                else
+                    @test true
+                end
+            end
+            if !ismissing(actual) && actual ∉ installed
+                @warn "Unexpectedly found julia version $actual when running `$c -v`. Only installed $installed in testing."
+            end
+        end
+    end
+end
 
-    @static if Sys.iswindows()
-        @test_skip "interactive windows installer"
-    else
-        @test UpdateJulia.version_of("julia-1.4") == v"1.4.2"
-        @test UpdateJulia.version_of("julia-1.2") == v"1.2.0"
-        @test UpdateJulia.version_of("julia-1.0.4") == v"1.0.4"
-        @test UpdateJulia.version_of("julia-1.0.0") == v"1.0.0"
-        @test UpdateJulia.version_of("julia-1.0") == v10_latest
+if ("CI" => "true") ∈ ENV
+    printstyled("WARNING: this will both install out of date julia versions and change what the command `julia` points to.\n", color=Base.warn_color())
+    printstyled("If all goes well, it will finish with the latest stable version of julia installed.\n", color=Base.warn_color())
+
+    @testset "curated tests" begin
+        # TODO add any failing tests from the random matrix here so that we have a readable,
+        # reproducible, and efficient list of tests to start with.
+        @test update_julia() == UpdateJulia.latest()
     end
 
-    @test UpdateJulia.version_of("julia-1.5.1") == v"1.5.1"
-    @test UpdateJulia.version_of("julia-1.5.2") == v"1.5.2"
-    @test UpdateJulia.version_of("julia-1.5.3") == v"1.5.3"
-    @test UpdateJulia.version_of("julia-1.7.0-rc1") == v"1.7.0-rc1"
-    @test UpdateJulia.version_of("julia-1.7.0-rc3") == v"1.7.0-rc3"
-    @test UpdateJulia.version_of("julia-1.5") ∈ [v"1.5.1", v"1.5.3"] # Could be 1.5.1 or 1.5.3 depending on path order b.c. of coexisting systemwide and user installations
```

Comment:
TODO accept this limitation in the matrix

Reply:
We just don't test systemwide in the matrix now... It works, but is slightly suboptimal. At least we test it directly in curated, though, which is sufficient.